src/osd/OSDMap.cc

```diff
@@ -5331,6 +5393,113 @@ OSDMap::candidates_t OSDMap::build_candidates(
   return candidates;
 }
 
+float OSDMap::calc_wl_balance_score(CephContext *cct, int64_t pool_id,
+				    wl_balance_info_t *p_wlbmi) const
+{
+  ldout(cct,20) << __func__ << " pool " << get_pool_name(pool_id) << dendl;
+
+  ceph_assert(pools.count(pool_id));
+  const pg_pool_t* pool = get_pg_pool(pool_id);
+  auto num_pgs = pool->get_pg_num();
+
+  map<uint64_t,set<pg_t>> pgs_by_osd;
+  map<uint64_t,set<pg_t>> prim_pgs_by_osd;
+  map<uint64_t,set<pg_t>> acting_prims_by_osd;
+
+  pgs_by_osd = get_pgs_by_osd(cct, pool_id, &prim_pgs_by_osd, &acting_prims_by_osd);
+
+  ldout(cct,30) << __func__ << " Primaries for pool: "
+		<< prim_pgs_by_osd << dendl;
+  for (auto& [osd,pgs] : prim_pgs_by_osd) {
+    ldout(cct,20) << __func__ << " Pool " << pool_id << " OSD." << osd
+                  << " has " << pgs.size() << " primary PGs" << dendl;
+  }
+
+  auto num_osds = pgs_by_osd.size();
+
+  float avg_prims_per_osd = (float)num_pgs / (float)num_osds;
+  uint64_t max_prims_per_osd = 0;
+  uint64_t max_acting_prims_per_osd = 0;
+  //
+  //FIXME: This is a first cut - to be improved in future versions by looking
+  //       at the OSD weights. (look at crush weight as well)
+  //
+  float prim_affinity_sum = 0.;
+  float total_osd_weight = 0.;
+  float total_weighted_prim_affinity = 0.;
+
+  map<int,float> osds_crush_weight;
+  // Set up the OSDMap
+  OSDMap tmp_osd_map;
+  tmp_osd_map.deepish_copy_from(*this);
+  int ruleno = pools.at(pool_id).get_crush_rule();
+  tmp_osd_map.crush->get_rule_weight_osd_map(ruleno, &osds_crush_weight);
+  ldout(cct,20) << __func__ << " pool " << pool_id
+                << " ruleno " << ruleno
+                << " weight-map " << osds_crush_weight
+                << dendl;
+  for (auto [osd, oweight] : osds_crush_weight) {
+    total_osd_weight += oweight;
+    if (osds_crush_weight.count(osd)) {
+      total_weighted_prim_affinity += osds_crush_weight.at(osd) * get_primary_affinityf(osd);
+    }
```

Comment:
Why is this "if" check necessary? If we're iterating over `osds_crush_weight` with a for loop, I don't see a reason why any OSD we need in line 5444 might not be in the `osds_crush_weight` map.

Reply:
You also already have `oweight`, so you don't need `osds_crush_weight.at(osd)`.